src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,291 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.IntroduceUsingStatement
+{
+    internal abstract class AbstractIntroduceUsingStatementCodeRefactoringProvider<TStatementSyntax, TLocalDeclarationSyntax, TBlockSyntax> : CodeRefactoringProvider
+        where TStatementSyntax : SyntaxNode
+        where TLocalDeclarationSyntax : TStatementSyntax
+        where TBlockSyntax : SyntaxNode
+    {
+        protected abstract SyntaxList<TStatementSyntax> GetStatements(TBlockSyntax blockSyntax);
+
+        protected abstract TBlockSyntax WithStatements(TBlockSyntax blockSyntax, SyntaxList<TStatementSyntax> statements);
+
+        protected abstract TStatementSyntax CreateUsingStatement(TLocalDeclarationSyntax declarationStatement, SyntaxTriviaList sameLineTrivia, SyntaxList<TStatementSyntax> statementsToSurround);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+
+            var (declarationSyntax, _) =
+                await FindDisposableLocalDeclaration(document, span, context.CancellationToken).ConfigureAwait(false);
+
+            if (declarationSyntax != null)
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    FeaturesResources.Introduce_using_statement,
```

Comment:
note: thsi will need a VB specific version when you do VB.  This resource should move to C# since 'using' is C# specific, and 'Using' woud be correct for VB.

Reply:
Still applies.

Reply:
Still applies, even though I'm not planning to do VB?

Reply:
We could take this as C# only for now and add a VB version based on demand.